TypeScript: migrate jest-console package to TS

kushagra-goyal-14 opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- Link this PR to its associated issue with an appropriate keyword: Closes, See, Follow up to, etc. -->
Part of https://github.com/WordPress/gutenberg/issues/67691

<!-- In a few words, what is the PR actually doing? -->
Migrating the jest-console package to TypeScript.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->
To enhance DX and add type safety.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->
By porting the code and tests to TypeScript.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
Typecheck and unit tests.

kushagra-goyal-14 committed:
refactor: migrate jest-console package from js to ts for type safety

kushagra-goyal-14 committed:
fix: fix unit tests

- remove js test file
- add spaces and fix docs

kushagra-goyal-14 committed:
Merge branch 'trunk' into try/migrate-jest-console-ts

kushagra-goyal-14 committed:
Merge branch 'trunk' into try/migrate-jest-console-ts

kushagra-goyal-14 committed:
Merge branch 'trunk' into try/migrate-jest-console-ts

manzoorwanijk commented on packages/jest-console/src/test/index.test.ts:
```diff
@@ -58,15 +58,15 @@ describe( 'jest-console', () => {
-			expect( spy.assertionsNumber ).toBe( 0 );
+			expect( ( spy as any ).assertionsNumber ).toBe( 0 );
 
 			console[ methodName ]( message );
 
 			expect( console )[ matcherName ]();
-			expect( spy.assertionsNumber ).toBe( 1 );
+			expect( ( spy as any ).assertionsNumber ).toBe( 1 );
 
 			expect( console )[ matcherNameWith ]( message );
-			expect( spy.assertionsNumber ).toBe( 2 );
+			expect( ( spy as any ).assertionsNumber ).toBe( 2 );
```
Why do we need these `any` assertions here? I think those can be removed.

kushagra-goyal-14 replied:
Yes, I have removed those unnecessary types from the tests.

kushagra-goyal-14 committed:
fix: remove unnecessary type assertion for assertionsNumber in jest-console tests

kushagra-goyal-14 commented:
Hey @manzoorwanijk  ,

Thank you for the review.

I have updated the PR with the changes. Can you please re-review or ship the PR whenever possible.

manzoorwanijk reviewed (commented):
I still see some `as any` assertions but I am not sure if we can entirely get rid of all of those.

manzoorwanijk commented on packages/jest-console/src/matchers.ts:
```diff
@@ -7,8 +7,18 @@ import { matcherHint, printExpected, printReceived } from 'jest-matcher-utils';
+	calls: any[][];
+	matcherName: string;
+	methodName: string;
+	expected?: any[];
```
```suggestion
	calls: unknown[][];
	matcherName: string;
	methodName: string;
	expected?: unknown[];
```

manzoorwanijk commented on packages/jest-console/src/matchers.ts:
```diff
@@ -28,7 +38,12 @@ const createErrorMessage = ( spyInfo ) => {
+	spy: Mock,
+	matcherName: string,
+	methodName: string,
+	expected?: any[]
```
```suggestion
	expected?: unknown[]
```

manzoorwanijk commented on packages/jest-console/src/matchers.ts:
```diff
@@ -51,22 +66,21 @@ const createSpyInfo = ( spy, matcherName, methodName, expected ) => {
-	( matcherName, methodName ) => ( received ) => {
-		const spy = received[ methodName ];
+	( matcherName: string, methodName: string ) => ( received: any ) => {
+		const spy = received[ methodName ] as Mock;
```
We can get rid of `any` type and the assertion
```suggestion
	( matcherName: string, methodName: string ) =>
	( received: Record< string, Mock > ) => {
		const spy = received[ methodName ];
```

manzoorwanijk commented on packages/jest-console/src/matchers.ts:
```diff
@@ -51,22 +66,21 @@ const createSpyInfo = ( spy, matcherName, methodName, expected ) => {
+	methodName: string
+) =>
+	function ( received: any, ...expected: any[] ) {
+		const spy = received[ methodName ] as Mock;
```
```suggestion
	function ( received: Record< string, Mock >, ...expected: unknown[] ) {
		const spy = received[ methodName ];
```

manzoorwanijk commented on packages/jest-console/src/matchers.ts:
```diff
@@ -87,6 +101,6 @@ expect.extend(
 			};
 		},
-		{}
+		{} as Record< string, any >
```
Instead of this assertion here, I would pass this as the first argument to th `.reduce` generic

```ts
Record<
	string,
	( received: Record< string, Mock > ) => {
		pass: boolean;
		message: () => string;
	}
>
```

So, the `.reduce` call would look like this

```ts
Object.entries( supportedMatchers ).reduce<
	Record<
		string,
		( received: Record< string, Mock > ) => {
			pass: boolean;
			message: () => string;
		}
	>
>(/* Arguments here */);
```

manzoorwanijk reviewed (changes requested):
`unknown` is a safer type than `any` because it narrows the scope instead of expanding it.

So, let us use `unknown` where we don't really know the type.

We can also improve some `any` types with some properly defined types.

kushagra-goyal-14 committed:
feat: add ExtendedMock type and refactor matcher functions to use it

kushagra-goyal-14 committed:
fix: remove unnecessary comment about return type in matchers.ts

kushagra-goyal-14 commented:
I've implemented the suggested changes to improve type safety in the jest-console package. Created a dedicated types.ts file with proper interfaces, replaced any with unknown where appropriate, and used more specific types throughout the codebase.

manzoorwanijk reviewed (changes requested):
We also need to update the `types` field in `package.json` for the package to `build-types` but also handle the existing `types/index.d.ts`, which IMHO can be changed to `declarations.d.ts` and included via `files` field in tsconfig.json.

kushagra-goyal-14 committed:
feat: add TypeScript declarations for custom Jest matchers and update package configuration

kushagra-goyal-14 committed:
Merge branch 'trunk' into try/migrate-jest-console-ts

manzoorwanijk reviewed (approved):
Perfect. Thanks
